[View in Colaboratory](https://colab.research.google.com/github/kraftaa/Numpy-Pandas-Instructions/blob/master/exploringMLDataSetTrain.ipynb)

**Table of Contents**

- Exploring an ML Data Set


# Exploring ML Data Set

We'll work with the popular adult data set.The data set has been taken from UCI Machine Learning Repository. You can download the data from here. In this data set, the dependent variable is "target." It is a binary classification problem. We need to predict if the salary of a given person is less than or more than 50K.from google.colab import files
uploaded = files.upload()

In [6]:
from google.colab import files
uploaded = files.upload()



Saving train.csv to train (1).csv


In [8]:
import pandas as pd

#load the data
train  = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")

In [9]:
#check data set
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
age               32561 non-null int64
workclass         30725 non-null object
fnlwgt            32561 non-null int64
education         32561 non-null object
education.num     32561 non-null int64
marital.status    32561 non-null object
occupation        30718 non-null object
relationship      32561 non-null object
race              32561 non-null object
sex               32561 non-null object
capital.gain      32561 non-null int64
capital.loss      32561 non-null int64
hours.per.week    32561 non-null int64
native.country    31978 non-null object
target            32561 non-null object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


We see that, the train data has 32561 rows and 15 columns. Out of these 15 columns, 6 have integers classes and the rest have object (or character) classes. Similarly, we can check for test data. An alternative way of quickly checking rows and columns is

In [10]:
print ("The train data has",train.shape)
print ("The test data has",test.shape)

#Let have a glimpse of the data set
train.head()

The train data has (32561, 15)
The test data has (16281, 15)


age          workclass  fnlwgt   education  education.num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital.status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital.gain  capital.loss  hours.per.week  native.country  target  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

Now, let's check the missing values (if present) in this data.



In [11]:
nans = train.shape[0] - train.dropna().shape[0]
print ("%d rows have missing values in the train data" %nans)

nand = test.shape[0] - test.dropna().shape[0]
print ("%d rows have missing values in the test data" %nand)

2399 rows have missing values in the train data
1221 rows have missing values in the test data


We should be more curious to know which columns have missing values.



In [12]:
#only 3 columns have missing values
train.isnull().sum()

age                  0
workclass         1836
fnlwgt               0
education            0
education.num        0
marital.status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week       0
native.country     583
target               0
dtype: int64

Let's count the number of unique values from character variables.



In [13]:
cat = train.select_dtypes(include=['O'])
cat.apply(pd.Series.nunique)

workclass          8
education         16
marital.status     7
occupation        14
relationship       6
race               5
sex                2
native.country    41
target             2
dtype: int64

Since missing values are found in all 3 character variables, let's impute these missing values with their respective modes.



In [15]:
#Education
train.workclass.value_counts(sort=True)
train.workclass.fillna('Private',inplace=True)


#Occupation
train.occupation.value_counts(sort=True)
train.occupation.fillna('Prof-specialty',inplace=True)


#Native Country
train['native.country'].value_counts(sort=True)
train['native.country'].fillna('United-States',inplace=True)

Let's check again if there are any missing values left.



In [16]:
train.isnull().sum()


age               0
workclass         0
fnlwgt            0
education         0
education.num     0
marital.status    0
occupation        0
relationship      0
race              0
sex               0
capital.gain      0
capital.loss      0
hours.per.week    0
native.country    0
target            0
dtype: int64

Now, we'll check the target variable to investigate if this data is imbalanced or not.



In [17]:
#check proportion of target variable
train.target.value_counts()/train.shape[0]

 <=50K    0.75919
 >50K     0.24081
Name: target, dtype: float64

We see that 75% of the data set belongs to <=50K class. This means that even if we take a rough guess of target prediction as <=50K, we'll get 75% accuracy. Isn't that amazing? Let's create a cross tab of the target variable with education. With this, we'll try to understand the influence of education on the target variable.

In [18]:
pd.crosstab(train.education, train.target,margins=True)/train.shape[0]


target            <=50K      >50K       All
education                                  
 10th          0.026750  0.001904  0.028654
 11th          0.034243  0.001843  0.036086
 12th          0.012285  0.001013  0.013298
 1st-4th       0.004975  0.000184  0.005160
 5th-6th       0.009736  0.000491  0.010227
 7th-8th       0.018611  0.001228  0.019840
 9th           0.014957  0.000829  0.015786
 Assoc-acdm    0.024631  0.008139  0.032769
 Assoc-voc     0.031357  0.011087  0.042443
 Bachelors     0.096250  0.068210  0.164461
 Doctorate     0.003286  0.009398  0.012684
 HS-grad       0.271060  0.051442  0.322502
 Masters       0.023464  0.029452  0.052916
 Preschool     0.001566  0.000000  0.001566
 Prof-school   0.004699  0.012991  0.017690
 Some-college  0.181321  0.042597  0.223918
All            0.759190  0.240810  1.000000

We see that out of 75% people with <=50K salary, 27% people are high school graduates, which is correct as people with lower levels of education are expected to earn less. On the other hand, out of 25% people with >=50K salary, 6% are bachelors and 5% are high-school grads. Now, this pattern seems to be a matter of concern. That's why we'll have to consider more variables before coming to a conclusion.

If you've come this far, you might be curious to get a taste of building your first machine learning model. In the coming week we'll share an exclusive tutorial on machine learning in python. However, let's get a taste of it here.

We'll use the famous and formidable scikit learn library. Scikit learn accepts data in numeric format. Now, we'll have to convert the character variable into numeric. We'll use the labelencoder function.

In label encoding, each unique value of a variable gets assigned a number, i.e., let's say a variable color has four values ['red','green','blue','pink'].

Label encoding this variable will return output as: red = 2 green = 0 blue = 1 pink = 3

In [19]:
#load sklearn and encode all object type variables
from sklearn import preprocessing

for x in train.columns:
    if train[x].dtype == 'object':
        lbl = preprocessing.LabelEncoder()
        lbl.fit(list(train[x].values))
        train[x] = lbl.transform(list(train[x].values))

Let's check the changes applied to the data set.



In [20]:
train.head()


age  workclass  fnlwgt  education  education.num  marital.status  \
0   39          6   77516          9             13               4   
1   50          5   83311          9             13               2   
2   38          3  215646         11              9               0   
3   53          3  234721          1              7               2   
4   28          3  338409          9             13               2   

   occupation  relationship  race  sex  capital.gain  capital.loss  \
0           0             1     4    1          2174             0   
1           3             0     4    1             0             0   
2           5             1     4    1             0             0   
3           5             0     2    1             0             0   
4           9             5     2    0             0             0   

   hours.per.week  native.country  target  
0              40              38       0  
1              13              38       0  
2              40              38       0  
3              40              38       0  
4              40               4       0

As we can see, all the variables have been converted to numeric, including the target variable.



In [21]:
#<50K = 0 and >50K = 1
train.target.value_counts()

0    24720
1     7841
Name: target, dtype: int64